### ISSUE TITLE
BaseCloudEvent.getExtensionNames() breaks cloud event encapsulation

### ISSUE THREAD

**Issue by mskuba01** [cloudevents/sdk-java#581]
please replace

    @Override
    public Set<String> getExtensionNames() {
        return this.extensions.keySet();
    }

with something, like

    @Override
    public Set<String> getExtensionNames() {
        return Collections.unmodifiableSet (this.extensions.keySet());
    }

to avoid breaking encapsulation.

More details can be found here

https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html#keySet--
Returns a [Set](https://docs.oracle.com/javase/8/docs/api/java/util/Set.html) view of the keys contained in this map. The set is backed by the map, so changes to the map are reflected in the set, and vice-versa.

**Comment by pierDipi** [comment]
Hi @mskuba01, thanks for the suggestion, feel free to propose a PR

**Comment by thedevhaider** [comment]
@mskuba01, Yeah this makes sense. The first snippet returns a modifiable keySet which breaks the encapsulation. @pierDipi  can you please assign this to me?
